core/services/ocr2/plugins/ocr2keeper/evm/registry.go

```diff
@@ -0,0 +1,742 @@
+	defer cancel()
+
+	idMap := make(map[int64]activeUpkeep)
+
+	// get active upkeep ids from contract
+	ids, err := r.getLatestIDsFromContract(startupCtx)
+	if err != nil {
+		return fmt.Errorf("failed to get ids from contract: %s", err)
+	}
+
+	var offset int
+	for offset < len(ids) {
+		batch := FetchUpkeepConfigBatchSize
+		if len(ids)-offset < batch {
+			batch = len(ids) - offset
+		}
+
+		actives, err := r.getUpkeepConfigs(startupCtx, ids[offset:offset+batch])
+		if err != nil {
+			return fmt.Errorf("failed to get configs for id batch (length '%d'): %s", batch, err)
+		}
+
+		for _, active := range actives {
+			idMap[active.ID.Int64()] = active
+		}
+
+		offset += batch
+	}
+
+	r.mu.Lock()
+	r.active = idMap
+	r.mu.Unlock()
+
+	return nil
+}
+
+func (r *EvmRegistry) pollLogs() error {
+	var latest int64
+	var end int64
+	var err error
+
+	if end, err = r.poller.LatestBlock(); err != nil {
+		return fmt.Errorf("%w: %s", ErrHeadNotAvailable, err)
+	}
+
+	r.mu.Lock()
+	latest = r.lastPollBlock
+	r.lastPollBlock = end
+	r.mu.Unlock()
+
+	// if start and end are the same, no polling needs to be done
+	if latest == 0 || latest == end {
+		return nil
+	}
+
+	{
+		var logs []logpoller.Log
+
+		if logs, err = r.poller.LogsWithSigs(
+			end-logEventLookback,
+			end,
+			upkeepStateEvents,
+			r.addr,
+			pg.WithParentCtx(r.ctx),
+		); err != nil {
+			return fmt.Errorf("%w: %s", ErrLogReadFailure, err)
+		}
+
+		for _, log := range logs {
+			r.chLog <- log
+		}
+	}
+
+	return nil
+}
+
+func (r *EvmRegistry) registerEvents(addr common.Address) error {
+	// Add log filters for the log poller so that it can poll and find the logs that
+	// we need
+	filterID, err := r.poller.RegisterFilter(logpoller.Filter{
+		EventSigs: append(upkeepStateEvents, upkeepActiveEvents...),
+		Addresses: []common.Address{addr},
+	})
+	if err != nil {
+		r.mu.Lock()
+		r.filterID = filterID
+		r.mu.Unlock()
+	}
+	return err
+}
+
+func (r *EvmRegistry) processUpkeepStateLog(l logpoller.Log) error {
+
+	hash := l.TxHash.String()
+	if _, ok := r.txHashes[hash]; ok {
+		return nil
+	}
+	r.txHashes[hash] = true
+
+	rawLog := l.ToGethLog()
+	abilog, err := r.registry.ParseLog(rawLog)
+	if err != nil {
+		return err
+	}
+
+	r.lggr.Debugf("log detected for event %s in transaction %s", abilog.Topic(), hash)
+
+	switch l := abilog.(type) {
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepRegistered: // adds new upkeep id to registry
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepReceived: // adds multiple new upkeep ids to registry
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepUnpaused:
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepGasLimitSet:
+		r.addToActive(l.Id)
+	}
+
+	return nil
+}
+
+func (r *EvmRegistry) addToActive(id *big.Int) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, ok := r.active[id.Int64()]; !ok {
```

Comment:
upkeep IDs can be 256 bit IDs, how does this map work?

Reply:
That might be safer